crypto/chacha/asm/chacha-armv8-sve.pl

```diff
@@ -0,0 +1,1008 @@
+	ldp	$tmpw2,$tmpw3,[$state,#8]
+	dup	$xt0.s,$tmpw0
+	dup	$xt1.s,$tmpw1
+	dup	$xt2.s,$tmpw2
+	dup	$xt3.s,$tmpw3
+	ldp	$tmpw0,$tmpw1,[$state,#16]
+	ldp	$tmpw2,$tmpw3,[$state,#24]
+	add	@mx[0].s,@mx[0].s,$xt0.s
+	add	@mx[1].s,@mx[1].s,$xt1.s
+	add	@mx[2].s,@mx[2].s,$xt2.s
+	add	@mx[3].s,@mx[3].s,$xt3.s
+	dup	$xt4.s,$tmpw0
+	dup	$xt5.s,$tmpw1
+	dup	$xt6.s,$tmpw2
+	dup	$xt7.s,$tmpw3
+	ldp	$tmpw0,$tmpw1,[$state,#32]
+	ldp	$tmpw2,$tmpw3,[$state,#40]
+	add	@mx[4].s,@mx[4].s,$xt4.s
+	add	@mx[5].s,@mx[5].s,$xt5.s
+	add	@mx[6].s,@mx[6].s,$xt6.s
+	add	@mx[7].s,@mx[7].s,$xt7.s
+	dup	$xt0.s,$tmpw0
+	dup	$xt1.s,$tmpw1
+	dup	$xt2.s,$tmpw2
+	dup	$xt3.s,$tmpw3
+	ldp	$tmpw0,$tmpw1,[$state,#48]
+	ldp	$tmpw2,$tmpw3,[$state,#56]
+	index	$xt4.s,$ictr,1
+	add	@mx[8].s,@mx[8].s,$xt0.s
+	add	@mx[9].s,@mx[9].s,$xt1.s
+	add	@mx[10].s,@mx[10].s,$xt2.s
+	add	@mx[11].s,@mx[11].s,$xt3.s
+	dup	$xt5.s,$tmpw1
+	dup	$xt6.s,$tmpw2
+	dup	$xt7.s,$tmpw3
+	add	@mx[12].s,@mx[12].s,$xt4.s
+	add	@mx[13].s,@mx[13].s,$xt5.s
+	add	@mx[14].s,@mx[14].s,$xt6.s
+	add	@mx[15].s,@mx[15].s,$xt7.s
+___
+}
+
+sub SVE_TRANSFORMS() {
+	&add_states();
+	&transpose($xa0,$xb0,$xc0,$xd0);
+	&transpose($xa1,$xb1,$xc1,$xd1);
+	&transpose($xa2,$xb2,$xc2,$xd2);
+	&transpose($xa3,$xb3,$xc3,$xd3);
+$code.=<<___;
+	cmp	$blocks,$veclen
+	b.ge	10f
+	lsl	$dlen,$blocks,4
+	whilelo	p1.s,xzr,$dlen
+	cntp	$rsize,p1,p1.s
+___
+	&transpose($xa0,$xa1,$xa2,$xa3);
+	&transpose($xb0,$xb1,$xb2,$xb3);
+	&transpose($xc0,$xc1,$xc2,$xc3);
+	&transpose($xd0,$xd1,$xd2,$xd3);
+	&output_incr($xa0,$xa1,$xa2,$xa3);
+	&output_incr($xb0,$xb1,$xb2,$xb3);
+	&output_incr($xc0,$xc1,$xc2,$xc3);
+	&output_incr($xd0,$xd1,$xd2,$xd3);
+$code.=<<___;
+	add	$ctr,$ctr,$blocks
+	b	100f
+10:
+	ptrues p1.s,ALL
+___
+	&load_8x($xt0,$xt1,$xt2,$xt3,$xt4,$xt5,$xt6,$xt7);
+	&transpose($xa0,$xa1,$xa2,$xa3);
+	&transpose($xb0,$xb1,$xb2,$xb3);
+$code.=<<___;
+	eor	$xa0.d,$xa0.d,$xt0.d
+	eor	$xa1.d,$xa1.d,$xt1.d
+	eor	$xa2.d,$xa2.d,$xt2.d
+	eor	$xa3.d,$xa3.d,$xt3.d
+	eor	$xb0.d,$xb0.d,$xt4.d
+	eor	$xb1.d,$xb1.d,$xt5.d
+	eor	$xb2.d,$xb2.d,$xt6.d
+	eor	$xb3.d,$xb3.d,$xt7.d
+___
+	&transpose($xc0,$xc1,$xc2,$xc3);
+	&store_8x($xa0,$xa1,$xa2,$xa3,$xb0,$xb1,$xb2,$xb3);
+	&load_8x($xt0,$xt1,$xt2,$xt3,$xt4,$xt5,$xt6,$xt7);
+	&transpose($xd0,$xd1,$xd2,$xd3);
+$code.=<<___;
+	eor	$xc0.d,$xc0.d,$xt0.d
+	eor	$xc1.d,$xc1.d,$xt1.d
+	eor	$xc2.d,$xc2.d,$xt2.d
+	eor	$xc3.d,$xc3.d,$xt3.d
+	eor	$xd0.d,$xd0.d,$xt4.d
+	eor	$xd1.d,$xd1.d,$xt5.d
+	eor	$xd2.d,$xd2.d,$xt6.d
+	eor	$xd3.d,$xd3.d,$xt7.d
+___
+	&store_8x($xc0,$xc1,$xc2,$xc3,$xd0,$xd1,$xd2,$xd3);
+$code.=<<___;
+	add	$ctr,$ctr,$veclen
+100:
+___
+}
+}}}
+
+sub SVE_LOAD_STATES() {
+	my ($tmpw0,$tmpw1,$tmpw2,$tmpw3) = ("w9","w10","w11","w12");
+
+$code.=<<___;
+	index	@mx[12].s,$ictr,1
+	// FIXME following code are not functionally necessary
+	// but appear to be vital to perf test
+#if 1
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+#endif
```

Comment:
This seems like it probably shouldn't stay here for merging to OpenSSL unless you're saying it improves performance on real hardware, which would be very odd.

Reply:
yes, it does improve performance on real hardware, and yes, it is very odd.

Reply:
tried everything i could think of, this weird code still needed.

Reply:
Fair enough!